db/types.go

```diff
@@ -0,0 +1,173 @@
+
+	// ErrVersionDoesNotExist is returned when a DB version does not exist
+	ErrVersionDoesNotExist = errors.New("version does not exist")
+)
+
+// DBConnection represents a connection to a versioned database.
+// Records are accessed via transaction objects, and must be safe for concurrent creation
+// and read and write access.
+// Past versions are only accessible read-only.
+type DBConnection interface {
+	// Opens a read-only transaction at the current working version.
+	Reader() DBReader
+
+	// Opens a read-only transaction at a specified version.
+	// Returns ErrVersionDoesNotExist for invalid versions.
+	ReaderAt(uint64) (DBReader, error)
+
+	// Opens a read-write transaction at the current version.
+	ReadWriter() DBReadWriter
+
+	// Opens a write-only transaction at the current version.
+	Writer() DBWriter
+
+	// Returns all saved versions
+	Versions() VersionSet
+
+	// Saves the current version of the database and returns its ID.
+	// Waits for any pending Writer transactions to be discarded. TODO: is an error preferred?
+	SaveVersion() uint64
+
+	// Close closes the database connection.
+	Close() error
+
+	// Stats returns a map of property values for all keys and the size of the cache.
+	Stats() map[string]string
+}
+
+// TODO: rename to batch/transaction?
+// DBReader is a read-only transaction interface. It is safe for concurrent access.
+// Callers must call Discard when done with the transaction.
+//
+// Keys cannot be nil or empty, while values cannot be nil. Keys and values should be considered
+// read-only, both when returned and when given, and must be copied before they are modified.
+type DBReader interface {
+	// Get fetches the value of the given key, or nil if it does not exist.
+	// CONTRACT: key, value readonly []byte
+	Get([]byte) ([]byte, error)
+
+	// Has checks if a key exists.
+	// CONTRACT: key, value readonly []byte
+	Has(key []byte) (bool, error)
+
+	// Iterator returns an iterator over a domain of keys, in ascending order. The caller must call
+	// Close when done. End is exclusive, and start must be less than end. A nil start iterates
+	// from the first key, and a nil end iterates to the last key (inclusive). Empty keys are not
+	// valid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// CONTRACT: start, end readonly []byte
+	Iterator(start, end []byte) (Iterator, error)
+
+	// ReverseIterator returns an iterator over a domain of keys, in descending order. The caller
+	// must call Close when done. End is exclusive, and start must be less than end. A nil end
+	// iterates from the last key (inclusive), and a nil start iterates to the first key (inclusive).
+	// Empty keys are not valid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// CONTRACT: start, end readonly []byte
+	ReverseIterator(start, end []byte) (Iterator, error)
+
+	// Discards the transaction, invalidating any future operations on it.
+	Discard()
```

Comment:
q: why is discard defined in both interfaces?

Reply:
Both read- and read-write transactions need to be closed/destroyed in the backend implementations, so we need to expose it for both.